Waterfall CI, test upstream LLVM wasm backend

kripken opened the pull request:
This gets CI for the upstream LLVM wasm backend on Circle. It uses the emsdk to get the lkgr build from the waterfall (using the code @jgravelle-google added), then uses that for the LLVM, clang, and Binaryen binaries.

This uses Python3 and found a bunch of bugs there, fixes already in separate PRs, also included here. With those fixes, this PR includes the same tests currently run on the waterfall (binaryen2, wasmobj2, other) and they all pass.

Questions:
 * Overhead: this does add some more time to our builds, but I think it's necessary it if we're serious about moving to the wasm backend by default. However we may have some pain until we improve the test infrastructure.
 * Naming questions: "waterfall"? "upstream" (used in the emsdk)?

jgravelle-google commented on .circleci/config.yml:
```diff
@@ -312,6 +313,75 @@ jobs:
+            tar -xf master.tar.gz
+            cd emsdk-master
+            ./emsdk --notty update-tags
+            export WATERFALL_LKGR=`python -c "print [x.split('\"')[3] for x in open('upstream/lkgr.json').readlines() if '\"build\"' in x][0]"`
```
That's pretty un-ergonomic yeah

kripken replied:
I couldn't find a better way :) Is there?

jgravelle-google replied:
Not that I'm aware of! Moreso that's a comment on how clunky it is to get upstream-llvm via emsdk, and how automation-unfriendly it is.
In principle this should be handled in emsdk itself, say `./emsdk install upstream-lkgr-64bit`, but I couldn't figure out how to get it to do that easily.

kripken replied:
Hmm yeah, the emsdk should really support this. We do have `emsdk install latest`, so we should add something like `latest-upstream` perhaps.

jgravelle-google commented on .circleci/config.yml:
```diff
@@ -312,6 +313,75 @@ jobs:
+  test-waterfall-other:
+    <<: *test-defaults
+    environment:
+      - TEST_TARGET=other skip:other.test_native_link_error_message skip:other.test_emcc_v skip:other.test_llvm_lit
```
Should we make these skips be `@no_wasm_backend` annotations in test_other.py?

sbc100 replied:
Its not really the wasm backend that makes the lit test fail.. its more like it can't find the llvm sources, which is true also for an installed fastcmp. 

Seems like running lit tests from here is a little strange.. should really be done when building fastcmp itself, no?  Should we remove test_llvm_lit completely?

kripken replied:
It's a little odd, yeah, but the reason is that we don't build fastcomp on CI (we use a prebuilt binary). This gives us coverage despite that.

I'd say adding an annotation in the python files makes sense for that one. For the other two, the issues are CI related, similar to the fastcomp testing, I'll add a comment.

jgravelle-google commented on .circleci/config.yml:
```diff
@@ -312,6 +313,75 @@ jobs:
+            # persistent workspace and to avoid any confusion with that version
+            # we are trying to test.
+            rm -r ~/emsdk-master/upstream/$WATERFALL_LKGR/emscripten/
+            # Use Binaryen from the emsdk's waterfall (wasm backend can't use ports)
```
More-specifically we can't use Binaryen as a port, ports-in-general should still work

kripken replied:
Good point, this was very unclear, fixing.

jgravelle-google commented on .circleci/config.yml:
```diff
@@ -312,6 +313,75 @@ jobs:
     <<: *test-defaults
     environment:
       - TEST_TARGET=sanity
+  build-waterfall:
```
Bikeshedding what to use instead of "waterfall" here, I'd say either "backend", "wasm_backend", or "upstream"

kripken replied:
Sounds good, emsdk already has it that way anyhow, updating.

sbc100 commented on .circleci/config.yml:
```diff
@@ -312,6 +313,75 @@ jobs:
+            echo "JS_ENGINES = [NODE_JS]" >> ~/.emscripten
+            echo "==="
+            cat ~/.emscripten
+            echo "==="
```
Probably don't need this debugging in the final version?

jgravelle-google replied:
Might be useful enough to clear up any assumptions that my instinct says leave it.
Having more info than usual on CI makes debugging it much easier.
Not that I expect the `.emscripten` specifically to cause too much fuss, but in general.

kripken replied:
Fixing, thanks.

kripken replied:
Oops, just saw @jgravelle-google's comment after I wrote mine.

I agree in general that debugging can be useful on CI, but I admit this particular one was mainly useful in bringing it up, and likely won't help much later.

kripken commented:
Addressed review feedback.

A funny thing is that given the fastcomp path is waiting on new binaries to go green, the upstream one may be the only one green on this run ;)

Oh, another question: this pulls the lkgr version from the waterfall. Is there a risk of that being updated too quickly or too slowly and show errors here? My guess is its safer than the fastcomp path (which has breakage for every breaking binary build change).
